jetbrains-core/src/software/aws/toolkits/jetbrains/core/coroutines/scopes.kt

```diff
@@ -0,0 +1,85 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.coroutines
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.components.service
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Disposer
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import java.util.concurrent.CancellationException
+
+/**
+ * Coroutine scope that is tied to the full Application closing, or plugin unloading. Default to dispatching to background thread pool.
+ *
+ * Use this if the coroutine needs to live past a project being closed or across projects such as an Application Service
+ */
+fun applicationCoroutineScope(coroutineName: String): CoroutineScope =
+    PluginCoroutineScopeTracker.getInstance().applicationThreadPoolScope(coroutineName)
+
+/**
+ * Coroutine scope that is tied to a project closing, or plugin unloading. Default to dispatching to background thread pool.
+ *
+ * Use this if the coroutine needs to live past a UI being closed, or tied to a project's life cycle such as a Project Service.
+ */
+fun projectCoroutineScope(project: Project, coroutineName: String): CoroutineScope =
+    PluginCoroutineScopeTracker.getInstance(project).applicationThreadPoolScope(coroutineName)
+
+/**
+ * Coroutine scope that is tied to a disposable or a plugin unloading. Default to dispatching to background thread pool.
+ *
+ * Use this if the coroutine is tied to a UI such as a tool window, or a run configuration
+ *
+ * **Note: If a call lives past the closing of a UI such as kicking off a resource creation, use [projectCoroutineScope].
+ * Otherwise, the coroutine will be canceled when the UI is closed!**
+ */
+fun disposableCoroutineScope(disposable: Disposable, coroutineName: String): CoroutineScope =
```

Comment:
should we ban attaching to a project/application with this one?

Reply:
I debated that, maybe we should throw illegal argument so prevent that.